crypto/ct/ct_lib.c

```diff
@@ -0,0 +1,222 @@
+ *    acknowledgment:
+ *    "This product includes software developed by the OpenSSL Project
+ *    for use in the OpenSSL Toolkit (http://www.OpenSSL.org/)"
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE OpenSSL PROJECT ``AS IS'' AND ANY
+ * EXPRESSED OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE OpenSSL PROJECT OR
+ * ITS CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT
+ * NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
+ * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
+ * STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED
+ * OF THE POSSIBILITY OF SUCH DAMAGE.
+ * ====================================================================
+ *
+ * This product includes cryptographic software written by Eric Young
+ * (eay@cryptsoft.com).  This product includes software written by Tim
+ * Hudson (tjh@cryptsoft.com).
+ *
+ */
+
+#ifndef OPENSSL_NO_CT
+
+# include <limits.h>
+# include "internal/cryptlib.h"
+# include "../ssl/ssl_locl.h"
+# include "ct_locl.h"
+
+SCT *SCT_new(void)
+{
+    SCT *sct = OPENSSL_zalloc(sizeof(SCT));
+    if (sct == NULL) {
+        CTerr(CT_F_SCT_NEW, ERR_R_MALLOC_FAILURE);
+        return NULL;
+    }
+    sct->entry_type = UNSET_ENTRY;
+    sct->version = UNSET_VERSION;
+    return sct;
+}
+
+void sct_free_internal(SCT *sct)
+{
+    if (sct->logid)
+        OPENSSL_free(sct->logid);
+    if (sct->ext)
+        OPENSSL_free(sct->ext);
+    if (sct->sig)
+        OPENSSL_free(sct->sig);
+    if (sct->sct)
+        OPENSSL_free(sct->sct);
+}
+
+void SCT_free(SCT *sct)
+{
+    if (sct) {
+        sct_free_internal(sct);
+        OPENSSL_free(sct);
+    }
+}
+
+int SCT_set_version(SCT *sct, sct_version_t version)
+{
+    if (version != SCT_V1) {
+        CTerr(CT_F_SCT_SET_VERSION, CT_R_UNSUPPORTED_VERSION);
+        return 0;
+    }
+    sct->version = version;
+    return 1;
+}
+
+int SCT_set_log_entry_type(SCT *sct, log_entry_type_t entry_type)
+{
+    if (entry_type != X509_ENTRY && entry_type != PRECERT_ENTRY) {
+        CTerr(CT_F_SCT_SET_LOG_ENTRY_TYPE, CT_R_UNSUPPORTED_ENTRY_TYPE);
+        return 0;
+    }
+    sct->entry_type = entry_type;
+    return 1;
+}
+
+int SCT_set0_logid(SCT *sct, void *logid, size_t logidlen)
+{
+    /* Currently only SHA-256 allowed so length must be SCT_V1_HASHLEN */
+    if (logidlen != SCT_V1_HASHLEN) {
+        CTerr(CT_F_SCT_SET0_LOGID, CT_R_INVALID_LOGID_LENGTH);
+        return 0;
+    }
+    if (sct->logid)
+        OPENSSL_free(sct->logid);
+    sct->logid = logid;
+    sct->logidlen = logidlen;
+    return 1;
+}
+
+int SCT_set_timestamp(SCT *sct, uint64_t timestamp)
```

Comment:
This can be void, since it always succeeds.

Unless you want to leave the door open for future SCT versions that don't have a timestamp field? But given that an SCT is a Signed Certificate *Timestamp*, I find that... unlikely?